numpy/_core/src/multiarray/ctors.c

```diff
@@ -658,6 +658,19 @@ PyArray_NewFromDescr_int(
         return NULL;
     }
 
+    /* finalize the descriptor if the DType defines a finalization function */
+    finalize_descr_function *finalize =
+            NPY_DT_SLOTS(NPY_DTYPE(descr))->finalize_descr;
+    if (finalize != NULL && data == NULL) {
+        PyArray_Descr *finalize_descr = finalize(descr);
+        if (finalize_descr == NULL) {
+            Py_DECREF(descr);
+            return NULL;
+        }
+        /* may or may not be the same as descr */
+        Py_SETREF(descr, finalize_descr);
```

Comment:
```suggestion
        Py_SETREF(descr, finalize(descr));
        if (descr == NULL) {
            return NULL;
        }
```
SETREF already handles any dance necessary.  (Not that it matters, if you prefer, please just go ahead and merge as is.)